[AST] Introduce ArgumentList

hamishknight opened the pull request:
Introduce the `ArgumentList` type, which represents a set of call arguments for a function or subscript. This supersedes the use of tuple and paren exprs as argument lists, a relic from when tuple splatting was a language feature.

For now, paren and tuple types retain their `ParameterTypeFlags`, as both the constraint system and a small amount of diagnostic logic is still relying on converting argument lists into tuple or paren types, this will be removed in a follow-up. This then sets the stage for the removal of `InOutExpr` and `InOutType`.

I've split the commits up based on the part of the compiler they touch, please feel free to review the bits of the compiler you're most familiar with.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@nathawes Mind taking a look at the changes specifically in Formatting.cpp?

hamishknight commented:
@swift-ci please SourceKit stress test

slavapestov commented:
Nice work!

CodaFi reviewed (approved):
Taken a few passes over this and I haven't seen anything so far. Sema, AST, Parse, and SIL changes all LGTM. CS LGTM but would prefer @hborla and @xedin give their rubber stamp there.

hamishknight commented:
@swift-ci please test source compatibility Debug

nathawes reviewed (approved):
The Formatting.cpp changes LGTM. It's so great this is finally happening – thanks a lot!

hamishknight commented:
@swift-ci please test compiler performance

xedin commented on include/swift/Sema/ConstraintSystem.h:
```diff
@@ -2676,25 +2666,19 @@ class ConstraintSystem {
-  /// that locator.
-  llvm::DenseMap<ConstraintLocator *, ArgumentInfo> ArgumentInfos;
+  /// constructions) to the argument lists for the call to that locator.
+  llvm::DenseMap<ConstraintLocator *, ArgumentList *> ArgumentLists;
```
While we are at it I think we need to do change this to behave like `addNodeTypes` and record `ArgumentLists` in the solution as well. Since constraint generation for result builders happens during solving this list would effectively get rewritten multiple times with the out-of-scope `ArgumentList` objects just to keep them around until constraint system lifetime ends.

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -110,18 +110,9 @@ Expr *FailureDiagnostic::findParentExpr(const Expr *subExpr) const {
-  return getAsExpr(simplifyLocatorToAnchor(argListLoc));
+ArgumentList *
+FailureDiagnostic::getArgumentListFor(ConstraintLocator *locator) const {
+  return getConstraintSystem().getArgumentList(locator);
```
See my comment about `ArgumentLists` - It'd be great if we could extract this information for `Solution` instead because we can't reply on constraint system having solution applied always e.g. ambiguity cases.

xedin reviewed (approved):
Looks great! I have left a small comment about `ArgumentLists` that could be addressed separately.

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -4073,21 +4024,22 @@ FuncDecl *isOperator(const BinaryExpr *BE) {
 ///   completion(nil, MyError.Bad) // Result = [MyError.Bad], IsError = true
 /// }
 class HandlerResult {
```
Maybe we should just store the ArgList with the range used instead? This is fine as is for now though I think. It'd just be nice to avoid creating the vectors where we don't really need them.

hamishknight replied:
Hmm, IMO the vector isn't too bad as the number of elements should be small in most cases, plus it's nice being able to expose an `ArrayRef`, and it would also give us the flexibility to support non-contiguous argument ranges if we needed to, what do you think?

bnbarham replied:
Yeah, definitely wouldn't worry about it for now even if we do end up changing it. This is the most 1:1 change 👍

bnbarham reviewed (approved):
Awesome 😍! The refactoring changes look good to me (two comments but it's fine as is).

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -7072,17 +7027,17 @@ class AsyncConverter : private SourceEntityWalker {
 
     Expr *CallbackArg =
-        lookThroughFunctionConversionExpr(ArgList.ref()[HandlerDesc.Index]);
+        lookThroughFunctionConversionExpr(ArgList->getExpr(HandlerDesc.Index));
```
Ah, seems it's not quite the last use, it's also used a couple of lines above to check its `size()`

bnbarham replied:
Heh, missed that one - was starting from below there 😆. Sorry for the noise!

hamishknight commented:
Compiler performance numbers are just noise, Jenkins is sporadically crashing during some of the source compat suite runs. Previous performance runs didn't show any meaningful differences though, only a slim decrease in AST memory allocation.

benlangmuir commented on lib/IDE/SyntaxModel.cpp:
```diff
@@ -547,15 +522,31 @@ static bool shouldTreatAsSingleToken(const SyntaxStructureNode &Node,
+std::pair<bool, ArgumentList *>
+ModelASTWalker::walkToArgumentListPre(ArgumentList *ArgList) {
+  for (auto Arg : *ArgList) {
+    auto res = ArgumentInfo.insert({Arg.getExpr(), Arg});
```
I guess this is going to store 4 extra pointers for every argument expression in a source file, which might be kind of big.  WDYT of keeping a counter of active argument lists, decremented in `walkToArgumentList`, and if it hits 0 we call `ArgumentInfo.clear()`?  That avoids the cost of removing individual expressions but keeps the map from growing more than necessary.

hamishknight replied:
Yeah, makes sense, how does this look?

benlangmuir replied:
LGTM, thanks!

benlangmuir reviewed (approved):
🎉 lib/Index, lib/IDE, ASTWalker, and SourceKit all LGTM.  I made one suggestion in SyntaxModel, but it doesn't need to block the PR, I think.  I love how much horrible code this removed.

hamishknight commented on lib/IDE/SourceEntityWalker.cpp:
```diff
@@ -262,6 +265,25 @@ static SemaReferenceKind getReferenceKind(Expr *Parent, Expr *E) {
+  if (ArgList->isImplicit())
+    return {true, ArgList};
+
+  // FIXME: What about SubscriptExpr and KeyPathExpr arg labels?
```
Ah thanks, I missed that one when filing bugs for the other fix-mes

hamishknight commented:
@swift-ci please test

rintaro commented on lib/Parse/ParseStmt.cpp:
```diff
@@ -840,21 +840,14 @@ ParserResult<Stmt> Parser::parseStmtYield(SourceLoc tryLoc) {
+                           lpLoc, yieldElts, rpLoc, SyntaxKind::ExprList);
+    for (auto &elt : yieldElts) {
+      assert(elt.Label.empty());
+      assert(elt.LabelLoc.isInvalid());
```
(Unrelated to this PR)
Hmm, I just noticed the following code hits these assertion even with current `main`.
```swift
struct S {
  var property: String {
    _read {
      yield(x: "test")
    }
  }
}
```
Instead of skipping parsing labels in `parseExprList()`, we probably should diagnose parsed labels here. Or, we should not use `parseExprList`.

hamishknight replied:
Good catch! Looks like the logic in `parseList` is trying to handle this case, but isn't working as `SyntaxKind::ExprList` is being passed (though looking at the git history, this seems to have been changed to fix the syntax tree). I can take a look at this in a follow-up.

rintaro reviewed (approved):
lib/Parse, lib/IDE, SourceKit changes LGTM!

hamishknight commented:
Stress tester run uncovered a nasty bug in the ASTWalker logic, now fixed

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
@swift-ci please test Windows

beccadax commented on include/swift/AST/ArgumentList.h:
```diff
@@ -0,0 +1,649 @@
+      --*this;
+      return tmp;
+    }
+
+    iterator &operator+=(unsigned i) {
+      Idx += i;
+      return *this;
+    }
+    iterator operator+(unsigned i) const {
+      return iterator(ArgList, Idx + i);
+    }
+    iterator &operator-=(unsigned i) {
+      Idx -= i;
+      return *this;
+    }
+    iterator operator-(unsigned i) const {
+      return iterator(ArgList, Idx - i);
+    }
+
+    unsigned operator-(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx - rhs.Idx;
+    }
+    bool operator<(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx < rhs.Idx;
+    }
+    bool operator<=(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx <= rhs.Idx;
+    }
+    bool operator>(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx > rhs.Idx;
+    }
+    bool operator>=(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx >= rhs.Idx;
+    }
+  };
```
Is this basically just `llvm::indexed_accessor_iterator`? (I think this class is relatively new; there might be other places we could adopt it in Swift.)

hamishknight replied:
Ah neat! I'll go ahead and change it over to use that

beccadax commented on include/swift/AST/ArgumentList.h:
```diff
@@ -0,0 +1,649 @@
+  /// Note that in type-checked AST, this may return an index of a variadic
+  /// expansion which contains the trailing closure (and it may not necessarily
+  /// be the first argument of the expansion).
+  Optional<unsigned> getRawFirstTrailingClosureIndex() const {
```
I was under the impression that the word "raw" referred to the fact that we were storing it in this weird format where `NumArgs` means `None`, but here you're calling it `getRawFirstTrailingClosureIndex()` even though you're converting it from that format. So what does "raw" mean in this name?

hamishknight replied:
The intention behind using "raw" was to communicate that the index you get might not be quite what you're expecting for type-checked AST (e.g you shouldn't in general do `auto numTrailing = argList->size() - *argList->getRawFirstTrailingClosureIndex()`). I struggled to think of another name that would communicate this, do you have any suggestions?

Alternatively we could rename the stored property to something like `FirstTrailingClosureIndexBits` to avoid conflating the meaning of "raw" here.

beccadax replied:
Hmm. I'm trying to understand this…

* The comment mentions varargs; I think it's saying that the first trailing closure argument might be in a `VarargExpansionExpr`. Is that correct? If so, what are you supposed to do in this situation? Is the `VarargExpansionExpr` going to somehow communicate to you which element is the first trailing closure argument?

* Are there other situations where `getRawFirstTrailingClosureIndex()` is "wrong"?

hamishknight replied:
> The comment mentions varargs; I think it's saying that the first trailing closure argument might be in a `VarargExpansionExpr`. Is that correct?

Yup, that's correct

> If so, what are you supposed to do in this situation? Is the `VarargExpansionExpr` going to somehow communicate to you which element is the first trailing closure argument?

Yeah, unless you look at source ranges, you wouldn't be able to know. It could be that there's some valid use case from looking at the first argument that contains a trailing closure, but it seems unlikely. As said in my other comment, I'm starting to think the right answer here might be to completely forbid calling this on a type-checked argument list.

> Are there other situations where `getRawFirstTrailingClosureIndex()` is "wrong"?

I don't believe so, it should always point to the first argument that is trailing outside of the r-paren source range. Though there are still other footguns with it, e.g default arguments and regular arguments that have been re-ordered may appear after it.

beccadax replied:
It's probably worth asking, what's the purpose of recording which arguments are "trailing closure arguments"? If it's so that the typechecker knows which arguments should be typechecked specially (e.g. loose argument label matching), you could argue that the typechecker basically turns trailing closure arguments into normal arguments. Thus, any (successfully) typechecked argument list will inherently have no trailing closure arguments.

hamishknight replied:
As far as I can tell the trailing closure info serves two main purposes:

1. For the expression type-checker, including solving and diagnostics
2. For various IDE related logic

The former mostly operates on un-type-checked AST (with the exception of MiscDiagnostics), so should be able to query trailing closure info from the argument list with no problems. The latter mostly operates on type-checked AST, so should always go through `getOriginalArguments()`.

> you could argue that the typechecker basically turns trailing closure arguments into normal arguments. Thus, any (successfully) typechecked argument list will inherently have no trailing closure arguments.

I really like the cleanness of that model, though I do worry that code might accidentally query trailing closure info on a type-checked argument list and not get the answer it's expecting (as it probably should have queried `getOriginalArguments()` instead). So I feel like some kind of warning, or assertion failure if you call it on type-checked AST would be useful there.

hamishknight replied:
Updated to force users to query through `getOriginalArgs()` for type-checked argument lists, as I feel it's important to avoid accidentally querying the wrong trailing closure info, and it's what the IDE + MiscDiagnostic logic tends to want anyways, how does this look?

beccadax replied:
I like it!

beccadax commented on include/swift/AST/ArgumentList.h:
```diff
@@ -0,0 +1,649 @@
+  /// Note that in type-checked AST, this may return \c true for an index of a
+  /// variadic expansion which contains a trailing closure (and it may not
+  /// necessarily be the first argument of the expansion).
+  bool isRawTrailingClosureIndex(unsigned i) const {
```
`isTrailingClosureArgument()`?

hamishknight replied:
Hmm, I think calling out the fact that it's a "raw" index is important, as it's not what you may expect in type-checked AST, in particular it might return true for default arguments or arguments that have been re-ordered after the trailing closure.

Alternatively we may just want to completely ban calling this on a type-checked argument list, what do you think?

hamishknight replied:
With the original arguments update this becomes `isTrailingClosureIndex`. Personally I prefer that over `isTrailingClosureArgument`, as that sounds like it would expect an `Argument` parameter, but I'm happy to change it over

beccadax commented on include/swift/AST/ArgumentList.h:
```diff
@@ -0,0 +1,649 @@
+/// list. For a type-checked argument list, this strips default arguments,
+/// expands variadic arguments, and re-orders trailing closures such that they
+/// are as they appear at the call-site.
+class OriginalArguments final {
```
Typechecking constructs a new `ArgumentList` instead of mutating the original, right? At least in the ways that we would want to undo for `OriginalArguments`?

If so, it might be better to have the typechecked `ArgumentList` contain a pointer to the original `ArgumentList` (perhaps in a trailing object so it can be removed when it's not needed) and then have `getOriginalArguments()` return that pointer if present, or `this` otherwise. That way we wouldn't need two separate types for representing argument lists, and we wouldn't have to reverse-engineer the original arguments from the ones in the AST.

hamishknight replied:
Yeah we construct a new `ArgumentList` in CSApply. Storing the un-type-checked argument list as the original did cross my mind when implementing this, but I wasn't sure how it would interact with additional AST transforms done after solver re-writing, and wasn't convinced it was worth the extra word in storage. That being said, storing it in the trailing objects is a clever idea to cut down on that cost, and it may well be that going straight back to the pre-solver-re-written argument list is exactly the behavior we want anyway.

I'll toy around with this and see how it goes! It certainly would be a nice simplification :)

hamishknight replied:
Okay, storing the pre-rewritten argument list appears to work just fine and simplifies things nicely, switched over to using that, thanks for the suggestion!

beccadax reviewed (commented):
This is going to be great! I've only looked at ArgumentList.h so far, but here are some initial thoughts. Sorry if you've already addressed some of this previously.

hamishknight commented on include/swift/AST/ArgumentList.h:
```diff
@@ -0,0 +1,649 @@
+  SourceLoc LabelLoc;
+  Identifier Label;
+  Expr *ArgExpr;
+  // TODO: Store inout bit here.
```
Yup, and then we can kill `InOutExpr` :)

hamishknight commented on include/swift/AST/ArgumentList.h:
```diff
@@ -0,0 +1,649 @@
+        IsImplicit(isImplicit), HasLabels(hasLabels),
+        HasLabelLocs(hasLabelLocs) {
+    assert(NumArgs == numArgs && "Num args truncated");
+    if (auto idx = firstTrailingClosureIndex) {
+      assert(*idx < numArgs);
+      RawFirstTrailingClosureIndex = *idx;
+    } else {
+      RawFirstTrailingClosureIndex = numArgs;
+      assert(!HasMultipleTrailingClosures);
+    }
```
Hmm the if statement felt better to me at the time because of the additional logic, but I don't really feel strongly about it at all, I can change it over

hamishknight commented on include/swift/AST/ArgumentList.h:
```diff
@@ -0,0 +1,649 @@
+                              ArrayRef<Argument> args, SourceLoc rParenLoc,
+                              Optional<unsigned> firstTrailingClosureIndex,
+                              bool hasMultipleTrailingClosures, bool isImplicit,
+                              Optional<AllocationArena> arena = None);
```
I didn't want to pull in `ASTContext.h` just for `AllocationArena`, and it didn't really feel like it was worth extracting into its own header – but with your `ASTAllocated` change we can totally change this.

Also TIL about `OptionalEnum`!

hamishknight commented on include/swift/AST/ArgumentList.h:
```diff
@@ -0,0 +1,649 @@
+  }
+
+  /// Retrieve the argument at a given index.
+  Argument get(unsigned idx) const {
```
`get` is used in most cases, but `operator []` is also required for template substitution for adaptors such as `indices`

hamishknight committed:
[AST] Adopt ArgumentList

Switch out the representation of argument lists
in various AST nodes with ArgumentList.

hamishknight committed:
[AST] Introduce ArgumentList

Introduce the ArgumentList type, which represents
a set of call arguments for a function or
subscript. This will supersede the use of tuple
and paren exprs as argument lists.

hamishknight committed:
[CS] Tighten up check for unresolved member chain

With the removal of TupleExpr and ParenExpr argument
lists, it's no longer sufficient to check whether
a given parent expr has a possible chain sub-expr,
as it might be a child that's unrelated to the chain
under consideration. For example, for a chain that's
an argument to a function call, we don't want to
walk up and consider the call expr to be an extension
of the chain just because it has a function expr.

Tighten up the check by comparing whether the given
sub-expr of a possible chain parent matches the
current expr in the chain.

hamishknight committed:
[Parser] Adopt ArgumentList

Split up the expr list parsing members such that
there are separate entry points for tuple and
argument list parsing, and start using the argument
list parsing member for call and subscripts.

hamishknight committed:
[CS] Adopt ArgumentList

- Explicitly limit favoring logic to only handle
unary args, this seems to have always been the
case, but needs to be handled explicitly now that
argument lists aren't exprs
- Update the ConstraintLocator simplification to
handle argument lists
- Store a mapping of locators to argument lists
in the constraint system
- Abstract more logic into a getArgumentLocator
method which retrieves an argument-to-param locator
from an argument anchor expr

hamishknight committed:
[MiscDiagnostics] Adopt ArgumentList

hamishknight committed:
[CodeSynthesis] Adopt ArgumentList

Most of this should be fairly mechanical, the
changes in PlaygroundTransform are a little more
involved though.

hamishknight committed:
[IDE] Adopt ArgumentList

hamishknight committed:
[Sema] Adopt ArgumentList

hamishknight committed:
[SIL] Adopt ArgumentList

hamishknight committed:
[SyntaxModel] Clear ArgumentInfo when outside of an argument list

Track the depth and clear out the map when it
reaches 0 to save memory.

hamishknight committed:
[ASTWalker] Fix ArgumentList walking bug

We were accidentally continuing to walk the old
argument list even if the caller had returned a
new one.

hamishknight committed:
Update for review feedback

- Remove OriginalArguments in favor of storing the
pre-rewritten argument list, which simplifies things
nicely
- Adopt llvm::indexed_accessor_iterator

hamishknight committed:
[Sema] Update getMatchVarIfInPatternMatch

hamishknight commented:
@ahoppen Mind taking a look at a1f127fa87f2a41b082d0f0d4424c6ec24ae43c3?

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test compiler performance

hamishknight commented:
@swift-ci please SourceKit stress test

hamishknight commented:
Stress tester result is a UPASS:

```
05:31:20 SourceKit Stress Tester summary:
05:31:20   failed underlying source compatibility build
05:31:20       > treat this as a source compatibility failure
05:31:20   4 XFails not processed
05:31:20   0 unexpected stress tester failures
05:31:20   135 expected stress tester failures tracked by 17 issues
```

The source compat failures are already known

ahoppen reviewed (approved):
a1f127f LGTM.

hamishknight commented:
🚢
